libraries/apollo-runtime-java/src/main/java/com/apollographql/apollo3/runtime/java/network/http/internal/BatchingHttpInterceptor.java

```diff
@@ -0,0 +1,256 @@
+  private final long batchIntervalMillis;
+  private final int maxBatchSize;
+  private final boolean exposeErrorBody;
+
+  private @NotNull HttpInterceptorChain interceptorChain;
+  private final List<PendingRequest> pendingRequests = new ArrayList<>();
+  private final ExecutorService executor = Executors.newSingleThreadExecutor();
+
+  public BatchingHttpInterceptor(long batchIntervalMillis, int maxBatchSize, boolean exposeErrorBody) {
+    this.batchIntervalMillis = batchIntervalMillis;
+    this.maxBatchSize = maxBatchSize;
+    this.exposeErrorBody = exposeErrorBody;
+  }
+
+  private void scheduleExecutePendingRequests() {
+    executor.execute(() -> {
+      try {
+        Thread.sleep(batchIntervalMillis);
+      } catch (InterruptedException ignored) {
+      }
+      executePendingRequests();
+    });
+  }
+
+  private static class PendingRequest {
+    private final HttpRequest request;
+    private final CountDownLatch latch = new CountDownLatch(1);
+    private HttpResponse response;
+    private ApolloNetworkException exception;
+
+    private PendingRequest(HttpRequest request) {
+      this.request = request;
+    }
+
+    private void setResponse(HttpResponse response) {
+      this.response = response;
+      latch.countDown();
+    }
+
+    private void setException(ApolloNetworkException exception) {
+      this.exception = exception;
+      latch.countDown();
+    }
+
+    private HttpResponse getResponse() throws ApolloNetworkException {
+      try {
+        latch.await();
+      } catch (InterruptedException ignored) {
+      }
+      if (exception != null) {
+        throw exception;
+      }
+      return response;
+    }
+  }
+
+  @Override public void intercept(@NotNull HttpRequest request, @NotNull HttpInterceptorChain chain, @NotNull HttpCallback callback) {
+    // Batching is enabled by default, unless explicitly disabled
+    String canBeBatchedHeader = HttpKt.get(request.getHeaders(), ExecutionOptions.CAN_BE_BATCHED);
+    boolean canBeBatched = canBeBatchedHeader == null || Boolean.parseBoolean(canBeBatchedHeader);
+    if (!canBeBatched) {
+      // Remove the CAN_BE_BATCHED header and forward directly
+      List<HttpHeader> headers = request.getHeaders().stream()
+          .filter(header -> !ExecutionOptions.CAN_BE_BATCHED.equals(header.getName()))
+          .collect(Collectors.toList());
+      chain.proceed(request.newBuilder().headers(headers).build(), callback);
+      return;
+    }
+
+    // Keep the chain for later
+    interceptorChain = chain;
+
+    PendingRequest pendingRequest = new PendingRequest(request);
+    int pendingRequestsSize;
+    synchronized (pendingRequests) {
+      pendingRequests.add(pendingRequest);
+      pendingRequestsSize = pendingRequests.size();
+    }
+    if (pendingRequestsSize >= maxBatchSize) {
+      executePendingRequests();
+    } else {
+      scheduleExecutePendingRequests();
+    }
+
+    try {
+      callback.onResponse(pendingRequest.getResponse());
```

Comment:
This blocks a thread. Could we pass `callback` to `PendingRequest` instead ?

Reply:
Good call! Simpler and works.